hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclCleaner.java

```diff
@@ -0,0 +1,156 @@
+    if (params != null && params.containsKey(HMaster.MASTER)) {
+      this.master = (HMaster) params.get(HMaster.MASTER);
+    }
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    super.setConf(conf);
+    userScanSnapshotEnabled = isUserScanSnapshotEnabled(conf);
+  }
+
+  @Override
+  protected boolean isFileDeletable(FileStatus fStat) {
+    // This plugin does not handle the file deletions, so return true by default
+    return true;
+  }
+
+  @Override
+  public boolean canDeleteEmptyDirectory(Path dir) {
+    if (userScanSnapshotEnabled) {
+      /*
+       * If user scan snapshot feature is enabled(see HBASE-21995), then when namespace or table
+       * exists, the archive namespace or table directories should not be deleted because the HDFS
+       * acls are set at these directories; the archive data directory should not be deleted because
+       * the HDFS acls of global permission is set at this directory.
+       */
+      return canDeleteEmptyArchiveDirectory(dir);
+    }
+    return true;
+  }
+
+  private boolean isUserScanSnapshotEnabled(Configuration conf) {
+    String masterCoprocessors = conf.get(CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY);
+    if (conf.getBoolean(SnapshotScannerHDFSAclHelper.USER_SCAN_SNAPSHOT_ENABLE, false)
+        && masterCoprocessors.contains(SnapshotScannerHDFSAclController.class.getName())
+        && masterCoprocessors.contains(AccessController.class.getName())) {
+      return true;
+    }
+    return false;
+  }
+
+  private boolean canDeleteEmptyArchiveDirectory(Path dir) {
+    if (isArchiveDataDir(dir)) {
+      return false;
+    } else if (isArchiveNamespaceDir(dir) && namespaceExists(dir.getName())) {
+      return false;
+    } else if (isArchiveTableDir(dir)) {
+      String name = dir.getName();
+      String namespace = dir.getParent().getName();
+      if (tableExists(TableName.valueOf(namespace, name))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @VisibleForTesting
+  static boolean isArchiveDataDir(Path path) {
+    if (path != null && path.getName().equals(HConstants.BASE_NAMESPACE_DIR)) {
+      Path parent = path.getParent();
+      return parent != null && parent.getName().equals(HConstants.HFILE_ARCHIVE_DIRECTORY);
+    }
+    return false;
+  }
+
+  @VisibleForTesting
+  static boolean isArchiveNamespaceDir(Path path) {
+    return path == null ? false : isArchiveDataDir(path.getParent());
+  }
+
+  @VisibleForTesting
+  static boolean isArchiveTableDir(Path path) {
+    return path == null ? false : isArchiveNamespaceDir(path.getParent());
+  }
+
+  private boolean namespaceExists(String namespace) {
+    try {
+      if (master != null) {
+        return master.listNamespaces().contains(namespace);
+      } else {
+        LOG.warn("Master is null when check if namespace {} exists", namespace);
+        return false;
+      }
+    } catch (IOException e) {
+      LOG.warn("Check if namespace {} exists error", namespace, e);
```

Comment:
Just throw the exception ? and handle in the outside  (CleanerChore#canDeleteEmptyDirectory )